shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -74,34 +74,44 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
   ) {
     def +=(b: Str -> TypeInfo): Unit = {
       env += b
-      if (quotedLvl > 0 && !qenv.contains(b._1)) {
+      if (inQuote && !qenv.contains(b._1)) {
         val tag = SkolemTag(freshVar(NoProv, N, nameHint = S(b._1))(lvl))(NoProv)
         println(s"Create skolem tag $tag for ${b._2} in quasiquote.")
         qenv += b._1 -> tag
       }
     }
     def ++=(bs: IterableOnce[Str -> TypeInfo]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str, qlvl: Int = 0): Opt[TypeInfo] =
-      if (qlvl === quotedLvl) env.get(name) orElse parent.dlof(_.get(name, qlvl))(N)
-      else parent.dlof(_.get(name, qlvl))(N)
+    def get(name: Str, quoted: Bool = false): Opt[TypeInfo] =
+      if (inQuote === quoted) env.get(name) orElse parent.dlof(_.get(name, quoted))(N)
+      else parent.dlof(_.get(name, quoted))(N)
     def getDecl(name: Str): Opt[NuDecl] = funDefs.get(name).map(_.decl) orElse parent.dlof(_.getDecl(name))(N)
     def getTopLevel(name: Str): Opt[TypeInfo] = (get(name), getDecl(name)) match {
       case (ty, S(fd: NuFunDef)) if (fd.outer.isEmpty) => ty
       case _ => N
     }
-    def qget(name: Str, qlvl: Int = quotedLvl): Opt[SkolemTag] =
-      if (qlvl === quotedLvl) qenv.get(name) orElse parent.dlof(_.qget(name, qlvl))(N)
-      else parent.dlof(_.qget(name, qlvl))(N)
+    def qget(name: Str, quoted: Bool = inQuote): Opt[SkolemTag] =
+      if (quoted === inQuote) qenv.get(name) orElse parent.dlof(_.qget(name, quoted))(N)
+      else parent.dlof(_.qget(name, quoted))(N)
     def getCtxTy: ST = fvars.foldLeft[ST](BotType)((res, ty) => res | ty)
     def wrapCode: Ls[(Str, TypeInfo)] = qenv.flatMap {
       case (name, tag) =>
-        get(name, quotedLvl) match {
+        get(name, inQuote) match {
           case S(VarSymbol(ty, _)) =>
             name -> VarSymbol(TypeRef(TypeName("Code"), ty :: tag :: Nil)(noProv), Var(name)) :: Nil
           case S(_: AbstractConstructor) | S(_: LazyTypeInfo) => die // * Abstract ctors and type defs are not allowed
           case N => Nil // * In the same quasiquote but not the same scope
         }
     }.toList
+    def unwrap[T](names: Ls[Str], f: () => T): T = { // * Revert ctx modification temporarily
```

Comment:
Can you please explain why this is needed?

Reply:
It would be much better to just add a flag to var symbols telling us whether they're quoted or not, rather than this imperative abomination.